gapir/cc/in_memory_resource_cache.cpp

```diff
@@ -23,174 +23,237 @@
+      mMemoryUse -= resRecord->second.first.getSize();
+
+      mResourceIndex.erase(resRecord->second.first.getID());
+      mResources.erase(resRecord);
+
+      // Get the data into the cache and return it.
+      return loadCacheMiss(res, target);
     }
-  }
-  // Whether we've grown or shrunk, the last block will always be free.
-  // Re-adjust the size so that it touches the first block.
-  last->size = (newSize - last->offset) + first->offset;
 
-  // If there's only one block remaining, it's free. Make sure it starts at 0.
-  if (last == first) {
-    last->offset = 0;
+    memcpy(target, &resRecord->second.second[0],
+           resRecord->second.first.getSize());
   }
 
-  mHead = last;  // Move head to the space.
-  mBufferSize = newSize;
+  // Update the bookkeeping for LRU to reflect this access.
+  auto newID = mIDGenerator++;
+  mResourceIndex[resRecord->second.first.getID()] = newID;
+  mResources[newID] = resRecord->second;
+  mResources.erase(resRecord);
+
+  // Note down the cache hit and return true
+  mCacheHits++;
   return true;
 }
 
-void InMemoryResourceCache::dump(FILE* out) {
-  Block* first = last()->next;
-  foreach_block(first, [&](Block* block) {
-    fprintf(out, (block == first) ? "┏━━━━━━━━━━━━━━━━" : "┳━━━━━━━━━━━━━━━━");
-  });
-  fprintf(out, "┓\n");
-  foreach_block(first, [&](Block* block) {
-    fprintf(out, "┃ offset: %6zu ", block->offset);
-  });
-  fprintf(out, "┃\n");
-  foreach_block(first, [&](Block* block) {
-    fprintf(out, "┃ size:   %6zu ", block->size);
-  });
-  fprintf(out, "┃\n");
-  foreach_block(first, [&](Block* block) {
-    if (block->isFree()) {
-      fprintf(out, "┃ free           ");
-    } else {
-      fprintf(out, "┃ id: %10.10s ", block->id.c_str());
-    }
-  });
-  fprintf(out, "┃\n");
-  foreach_block(first, [&](Block* block) {
-    fprintf(out, (block == mHead) ? "┃ head           " : "┃                ");
-  });
-  fprintf(out, "┃\n");
-  foreach_block(first, [&](Block* block) {
-    fprintf(out, (block == first) ? "┗━━━━━━━━━━━━━━━━" : "┻━━━━━━━━━━━━━━━━");
-  });
-  fprintf(out, "┛\n");
+size_t InMemoryResourceCache::totalCacheSize() const { return mMemoryLimit; }
+
+size_t InMemoryResourceCache::unusedSize() const {
+  return (size_t)(mMemoryLimit - mMemoryUse);
 }
 
-bool InMemoryResourceCache::putCache(const Resource& resource,
-                                     const void* data) {
-  if (resource.size > mBufferSize) {
-    return false;  // Wouldn't fit even if everything was evicted.
+bool InMemoryResourceCache::resize(size_t newSize) {
+  // Throw things out of the cache until we're below limit.
+  if (newSize < mMemoryUse) {
+    assert(evictLeastRecentlyUsed(mMemoryUse - newSize));
   }
 
-  // Merge mHead into next block(s) until it is big enough to hold our resource.
-  while (mHead->size < resource.size) {
-    mHead->size += mHead->next->size;
-    destroy(mHead->next);
+  return true;
+}
+
+void InMemoryResourceCache::dump(FILE* file) { assert(false); }
```

Comment:
It would be nice to keep our Dump() function working. It was useful to debug failures (e.g. when implementing split-replay).